### PR TITLE
[SHT_LLVM_BB_ADDR_MAP][llvm-readobj] Implements llvm-readobj handling for PGOAnalysisMap.

### PR DESCRIPTION
**Pull request by red1bluelost**
Adds raw printing of PGOAnalysisMap in llvm-readobj.

I'm leaving the fixme's for a later patch that will provide a 'pretty' printing for BBFreq and BrProb (i.e. relative frequencies and probabilities) that will apply to both llvm-readobj and llvm-objdump.

### CODE REVIEW

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test [hunk 1]
```diff
@@ -0,0 +1,221 @@
+## This test checks how we handle PGO Analysis Map in --bb-addr-map.
```
Comment (rlavaee) [inline llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test:1]: nit: I suggest rewording to "how llvm-readobj prints the PGO analysis map with the --bb-addr-map option."

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+## Fails on windows (https://github.com/llvm/llvm-project/issues/60013).
+# UNSUPPORTED: system-windows
```
Comment (rlavaee) [inline llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test:4]: Remove this.

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+# RUN: yaml2obj --docnum=1 %s -DBITS=64 -DADDR=0x999999999 -o %t1.v1.x64.o
+# RUN: llvm-readobj %t1.v1.x64.o --bb-addr-map 2>&1 | FileCheck %s -DADDR=0x999999999 -DFILE=%t1.v1.x64.o --check-prefix=CHECK
+
+## Check 32-bit:
+# RUN: yaml2obj --docnum=1 %s -DBITS=32 -o %t1.x32.o
+# RUN: llvm-readobj %t1.x32.o --bb-addr-map 2>&1 | FileCheck -DADDR=0x11111 %s -DFILE=%t1.x32.o --check-prefix=CHECK
+# RUN: llvm-readelf %t1.x32.o --bb-addr-map | FileCheck %s --check-prefix=GNU
+
+## Check that a malformed section can be handled.
+# RUN: yaml2obj --docnum=1 %s -DBITS=32 -DSIZE=6 -o %t2.o
```
Comment (rlavaee) [inline llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test:21]: Maybe truncate this at a larger offset which lies in the PGO analysis map data.

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+## Fails on windows (https://github.com/llvm/llvm-project/issues/60013).
+# UNSUPPORTED: system-windows
+
+## Check 64-bit:
+# RUN: yaml2obj --docnum=1 %s -DBITS=64 -DADDR=0x999999999 -o %t1.x64.o
+# RUN: llvm-readobj %t1.x64.o --bb-addr-map 2>&1 | FileCheck %s -DADDR=0x999999999 -DFILE=%t1.x64.o --check-prefix=CHECK
+# RUN: llvm-readelf %t1.x64.o --bb-addr-map | FileCheck %s --check-prefix=GNU
+
+## Check 64-bit:
+# RUN: yaml2obj --docnum=1 %s -DBITS=64 -DADDR=0x999999999 -o %t1.v1.x64.o
```
Comment (rlavaee) [inline llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test:12]: This looks not any different from the previous one (there is no version: 1).

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+# RUN: llvm-readobj %t1.v1.x64.o --bb-addr-map 2>&1 | FileCheck %s -DADDR=0x999999999 -DFILE=%t1.v1.x64.o --check-prefix=CHECK
+
+## Check 32-bit:
+# RUN: yaml2obj --docnum=1 %s -DBITS=32 -o %t1.x32.o
```
Comment (rlavaee) [inline llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test:16]: Remove all docnums as there is only a single doc in this test.

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test (hunk 2 above)
Comment (red1bluelost) [reply llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test]: Good catch, think I did this before the fix.

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test (hunk 3 above)
Comment (red1bluelost) [reply llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test]: Good idea, I've updated it so that it should fall in PGO data and checked the offset using objdump.

File: llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test (hunk 4 above)
Comment (red1bluelost) [reply llvm/test/tools/llvm-readobj/ELF/bb-addr-map-pgo-analysis-map.test]: Good catch, I think this is similarly the case in `bb-addr-map.test` which I adapted.

**Comment by red1bluelost** [comment]
I updated the metadata test to instead provide the wrong feature bit.

**Review by rlavaee — APPROVED** [review]
